proxy/tcpproxy/userspace.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpproxy
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+type tcpProxy struct {
+	l               net.Listener
+	monitorInterval time.Duration
+
+	mu      sync.Mutex // guards the following fields
+	remotes []*remote
+	next    int
+}
+
+type remote struct {
+	mu       sync.Mutex
+	addr     string
+	inactive bool
+}
+
+func (r *remote) inactivate() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.inactive = true
+}
+
+func (r *remote) tryReactivate() {
+	conn, err := net.Dial("tcp", r.addr)
+	if err == nil {
+		conn.Close()
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		r.inactive = false
+		return
+	}
+}
+
+func (r *remote) isActive() bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return !r.inactive
+}
+
+func (tp *tcpProxy) run() error {
+	for {
+		in, err := tp.l.Accept()
+		if err != nil {
+			return err
+		}
+
+		go tp.serve(in)
```

Comment:
@xiang90 Nvm. I thought of blocking goroutine like this:

```go
go wrap(forever())              // blocking
go func() { wrap(forever()) }() // non-blocking
```

but we don't wrap with anything, so don't matter.